processor/src/main/java/org/mapstruct/ap/internal/util/AbstractElementUtilsDecorator.java

```diff
@@ -34,20 +36,91 @@
 
 public abstract class AbstractElementUtilsDecorator implements ElementUtils {
 
+    private static final Method MODULE_OF_METHOD;
+    private static final Method GET_TYPE_ELEMENT_WITH_MODULE_METHOD;
+    private static final Method GET_PACKAGE_ELEMENT_WITH_MODULE_METHOD;
+
+    static {
+        Method moduleOfMethod;
+        Method getTypeElementWithModuleMethod;
+        Method getPackageElementWithModuleMethod;
+        try {
+            moduleOfMethod = ElementUtils.class.getMethod( "getModuleOf", Element.class );
+            Class<?> moduleElementClass = Class.forName( "javax.lang.model.element.ModuleElement" );
+            getTypeElementWithModuleMethod = ElementUtils.class.getMethod(
+                "getTypeElement",
+                moduleElementClass,
+                CharSequence.class
+            );
+            getPackageElementWithModuleMethod = ElementUtils.class.getMethod(
+                "getPackageElement",
+                moduleElementClass,
+                CharSequence.class
+            );
+
+        }
+        catch ( NoSuchMethodException | ClassNotFoundException e ) {
+            moduleOfMethod = null;
+            getTypeElementWithModuleMethod = null;
+            getPackageElementWithModuleMethod = null;
+        }
+
+        MODULE_OF_METHOD = moduleOfMethod;
+        GET_TYPE_ELEMENT_WITH_MODULE_METHOD = getTypeElementWithModuleMethod;
+        GET_PACKAGE_ELEMENT_WITH_MODULE_METHOD = getPackageElementWithModuleMethod;
+    }
+
     private final Elements delegate;
+    /**
+     * The module element when running with the module system.
+     * {@code null} otherwise.
+     */
+    private final Element moduleElement;
 
-    AbstractElementUtilsDecorator(ProcessingEnvironment processingEnv) {
+    AbstractElementUtilsDecorator(ProcessingEnvironment processingEnv, TypeElement mapperElement) {
         this.delegate = processingEnv.getElementUtils();
+        Element moduleElement;
+        if ( MODULE_OF_METHOD == null ) {
```

Comment:
this would then be a java version check. can be a static initialised boolean.

Reply:
We still need to invoke the method somehow, we can't compile with target 8 and invoke a Java 9 method.

Reply:
Are you sure about that? I thought that `target 8` means that it compiles target 8 compatible byte-code. I don't think the compiler actively checks the if a method (with a since java9) is present or not.  What would otherwise be the idea of having a `target 8` decoupled from a `source 8`? . Also, the classes in `javax.lang.model` are an extension to java and not the core right?

Reply:
OK I did some analysis. If we only use `source` and `target` we can make this work. However, if we switch to using the new `release` flag then there will be a compile error (same error as we would run the compilation with Java 8). This does mean that we have to build the processor module with at least java 9.

> Also, the classes in `javax.lang.model` are an extension to java and not the core right?

What do you mean an extension? They are part of the OpenJDK.

In any case if you are really against the reflection then I would prefer to use [Multi-Release JAR](https://openjdk.java.net/jeps/238). Keep in mind that this is not new artifact, we are still going to provide a single dependency. Since Java 9 you can provide different implementations of certain classes for different Java versions in the same JAR under `META-INF/versions/<javaVersion>`

Reply:
I mean.. records an lamda's are java core features (probably implemented in native  byte-code for that matter)

The javax stuff isn't, right?

Reply:
so my proposal would be: go to a higher java version but set the target bytecode to 8. I think you still would be able to call the methods if you run inside 9 java (you compiled to 9 level source). But it would also generate javac 8 compatible byte code since the target level is 8. But it could be that I'm all wrong about the sof course 😄 . Never hurts to try.

Reply:
I tried it already and we only use `source` and `target` we can achieve what we want. We'll need to make sure our tests setup runs properly. Let me push that here.